src/compiler/checker.ts

```diff
@@ -36923,7 +36923,7 @@ namespace ts {
                 // We only need `Awaited<T>` if `T` is a type variable that has no base constraint, or the base constraint of `T` is `any`, `unknown`, `{}`, `object`,
                 // or is promise-like.
                 if (baseConstraint ?
-                    baseConstraint.flags & TypeFlags.AnyOrUnknown || isEmptyObjectType(baseConstraint) || isThenableType(baseConstraint) :
+                    baseConstraint.flags & TypeFlags.AnyOrUnknown || isEmptyObjectType(baseConstraint) || someType(baseConstraint, isThenableType) :

```

Comment:
An alternative fix would be to move `someType` to `isThenableType`:
<details>
<summary>patch</summary>

```diff
diff --git a/src/compiler/checker.ts b/src/compiler/checker.ts
index 088ba325ec..89cfb6bcae 100644
--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -36884,8 +36884,10 @@ namespace ts {
                 return false;
             }
 
-            const thenFunction = getTypeOfPropertyOfType(type, "then" as __String);
-            return !!thenFunction && getSignaturesOfType(getTypeWithFacts(thenFunction, TypeFacts.NEUndefinedOrNull), SignatureKind.Call).length > 0;
+            return someType(type, t => {
+                const thenFunction = getTypeOfPropertyOfType(t, "then" as __String);
+                return !!thenFunction && getSignaturesOfType(getTypeWithFacts(thenFunction, TypeFacts.NEUndefinedOrNull), SignatureKind.Call).length > 0;
+            });
         }
 
         interface AwaitedTypeInstantiation extends Type {
@@ -36923,7 +36925,7 @@ namespace ts {
                 // We only need `Awaited<T>` if `T` is a type variable that has no base constraint, or the base constraint of `T` is `any`, `unknown`, `{}`, `object`,
                 // or is promise-like.
                 if (baseConstraint ?
-                    baseConstraint.flags & TypeFlags.AnyOrUnknown || isEmptyObjectType(baseConstraint) || someType(baseConstraint, isThenableType) :
+                    baseConstraint.flags & TypeFlags.AnyOrUnknown || isEmptyObjectType(baseConstraint) || isThenableType(baseConstraint) :
                     maybeTypeOfKind(type, TypeFlags.TypeVariable)) {
                     return true;
                 }
```
</details>

That would in turn impact the other call site to this util, [here](https://github.dev/microsoft/TypeScript/blob/ed7db78106664e52f00ec646ed6a05e079f433ca/src/compiler/checker.ts#L37082-L37108) and I'm not sure if that's desirable.